oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/OAuth2TokenGenerator.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2020-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization;
+
+import org.springframework.lang.Nullable;
+import org.springframework.security.oauth2.core.OAuth2Token;
+
+/**
+ * Implementations of this interface are responsible for generating an {@link OAuth2Token}
+ * using the attributes contained in the {@link OAuth2TokenContext}.
+ *
+ * @author Joe Grandja
+ * @since 0.2.3
+ * @see OAuth2Token
+ * @see OAuth2TokenContext
+ * @param <T> the type of the OAuth 2.0 Token
+ */
+@FunctionalInterface
+public interface OAuth2TokenGenerator<T extends OAuth2Token> {
```

Comment:
In the case of `JwtGenerator`, the generic type `T` is a `Jwt`. When there's an `OpaqueTokenGenerator` class, what will the generic type be? Is it just `OAuth2AccessToken`?

Reply:
I think it will be something like:

```java
public class OAuth2OpaqueToken extends AbstractOAuth2Token {
	private final Map<String, Object> claims;

	public OAuth2OpaqueToken(String tokenValue, Instant issuedAt, Instant expiresAt, @Nullable Map<String, Object> claims) {
		super(tokenValue, issuedAt, expiresAt);
		this.claims = Collections.unmodifiableMap(claims != null ? claims : Collections.emptyMap());
	}

}
```

This would be used for opaque access token (with `claims`) and potentially used for refresh tokens (with or without `claims`).